pymatgen/command_line/bader_caller.py

```diff
@@ -179,49 +192,71 @@ def __init__(
-
-                    # Shift the data so that the atomic charge density to the center for easier manipulation
-                    shifted_data = np.roll(chg.data["total"], shift, axis=(0, 1, 2))
-
-                    # Slices a central window from the data array
-                    def slice_from_center(data, x_width, y_width, z_width):
-                        x, y, z = data.shape
-                        start_x = x // 2 - (x_width // 2)
-                        start_y = y // 2 - (y_width // 2)
-                        start_z = z // 2 - (z_width // 2)
-                        return data[
-                            start_x : start_x + x_width,
-                            start_y : start_y + y_width,
-                            start_z : start_z + z_width,
-                        ]
-
-                    def find_encompassing_vol(data: np.ndarray):
-                        # Find the central encompassing volume which holds all the data within a precision
-                        total = np.sum(data)
-                        for idx in range(np.max(data.shape)):
-                            sliced_data = slice_from_center(data, idx, idx, idx)
-                            if total - np.sum(sliced_data) < 0.1:
-                                return sliced_data
-                        return None
-
-                    dct = {
-                        "data": find_encompassing_vol(shifted_data),
-                        "shift": shift,
-                        "dim": self.chgcar.dim,
-                    }
-                    atomic_densities.append(dct)
-                self.atomic_densities = atomic_densities
+                self._parse_atomic_densities()
+
+    @deprecated(
+        message=(
+            "parse_atomic_densities was deprecated on 2024-02-26 "
+            "and will be removed on 2025-02-26. See https://"
+            "github.com/materialsproject/pymatgen/issues/3652 for details."
+        )
+    )
+    def _parse_atomic_densities(self):
+        # Deprecation tracker
+        if datetime(2025, 2, 26) < datetime.now() and "CI" in os.environ:
+            raise RuntimeError("This method should have been removed, see #3656.")
+
+        def slice_from_center(data, x_width, y_width, z_width):
+            """Slices a central window from the data array."""
+            x, y, z = data.shape
+            start_x = x // 2 - (x_width // 2)
+            start_y = y // 2 - (y_width // 2)
+            start_z = z // 2 - (z_width // 2)
+            return data[
+                start_x : start_x + x_width,
+                start_y : start_y + y_width,
+                start_z : start_z + z_width,
+            ]
+
+        def find_encompassing_vol(data: np.ndarray):
+            """Find the central encompassing volume which
+            holds all the data within a precision.
+            """
+            total = np.sum(data)
+            for idx in range(np.max(data.shape)):
```

Comment:
Another note for my side: 

- This method `bader_caller` is overall slow, or the bottleneck could be called `bader` itself, and might need some profiling to identify the bottleneck,

- Try replacing linear search with bisection method to see if performance could be improved. However it's also important to note this should search for the **MINIMUM** encompassing volume, pure bisection might not be suitable.